lib/internal/worker.js

```diff
@@ -77,7 +119,171 @@ function setupPortReferencing(port, eventEmitter, eventName) {
+    super();
+    debug(`[${threadId}] create new worker`, filename, options);
+    if (typeof filename !== 'string') {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'filename',
+                                 'string', filename);
+    }
+
+    if (!options.eval && !path.isAbsolute(filename)) {
+      throw new errors.TypeError('ERR_WORKER_NEED_ABSOLUTE_PATH', filename);
+    }
+
+    const resourceLimits = {
+      maxSemiSpaceSize: options.maxSemiSpaceSize | 0,
+      maxOldSpaceSize: options.maxOldSpaceSize | 0
+    };
+
+    // Set up the C++ handle for the worker, as well as some internal wiring.
+    this[kHandle] = new WorkerImpl(resourceLimits);
+    this[kHandle].onexit = (code) => this[kOnExit](code);
+    this[kPort] = this[kHandle].messagePort;
+    this[kPort].on('message', (payload) => this.emit('message', payload));
+    this[kPort].on('flaggedMessage', (flag, payload) => {
+      this[kOnMessage](flag, payload);
+    });
+    this[kPort].start();
+    debug(`[${threadId}] created Worker with ID ${this.threadId}`);
+
+    this[kPort].postMessage([
+      filename, !!options.eval, options.workerData
+    ], undefined, kLoadScript);
+    // Actually start the new thread now that everything is in place.
+    this[kHandle].startThread();
+  }
+
+  [kOnExit](code) {
+    debug(`[${threadId}] hears end event for Worker ${this.threadId}`);
+    this[kDispose]();
+    this.emit('exit', code);
+    this.removeAllListeners();
+  }
+
+  [kOnCouldNotSerializeErr]() {
+    this.emit('error', new errors.Error('ERR_WORKER_UNSERIALIZABLE_ERROR'));
+  }
+
+  [kOnOutOfMemory](serialized) {
+    this.emit('error', new errors.Error('ERR_WORKER_OUT_OF_MEMORY'));
+  }
+
+  [kOnErrorMessage](serialized) {
+    // This is what is called for uncaught exceptions.
+    const error = deserializeError(serialized);
+    this.emit('error', error);
+  }
+
+  [kOnMessage](flag, payload) {
+    switch (flag) {
+      case MESSAGE_FLAG_NONE:
+        // This is a userland message.
+        return this.emit('message', payload);
+      case kUpAndRunning:
+        return this.emit('online');
+      case MESSAGE_FLAG_COULD_NOT_SERIALIZE_ERR:
+        return this[kOnCouldNotSerializeErr](payload);
+      case MESSAGE_FLAG_OUT_OF_MEMORY:
+        return this[kOnOutOfMemory](payload);
+      case MESSAGE_FLAG_ERROR_MESSAGE:
+        return this[kOnErrorMessage](payload);
+    }
+
+    assert.fail(`Unknown worker message flag ${flag}`);
+  }
+
+  [kDispose]() {
+    this[kHandle].onexit = null;
+    this[kHandle] = null;
+    this[kPort] = null;
+  }
+
+  postMessage(payload, transferList = undefined) {
+    if (this[kHandle] === null) return;
+
+    this[kPort].postMessage(payload, transferList);
+  }
+
+  terminate(callback) {
+    if (this[kHandle] === null) return;
+
+    debug(`[${threadId}] terminates Worker with ID ${this.threadId}`);
+
+    if (typeof callback !== 'undefined')
+      this.once('exit', (exitCode) => callback(null, exitCode));
```

Comment:
Why the null first argument if there's no error path?

Reply:
I guess just considering the error-first pattern as a standard in Node land? E.g. tools like `util.promisify` (or similar userland tools) rely on it

Reply:
There's plenty of similar examples of terminator functions binding a "callback" to an event which don't use the error-first style, like `net.close(cb)` and `stream.end(cb)`. IMO, *not* having the impossible error argument slot makes more sense, but I think it'd be better to hear what others think on this.

Reply:
Hm, yeah, but those don’t carry data like an exit code, right?

Reply:
Oh, yep. You're right. I was remembering the close event on `net.Socket` having an argument to indicate if it closed due to an exception, but that's only on client sockets, which don't have a `close(cb)`, only a `destroy(exception)` and `end(data, encoding)`.